### PR TITLE
Refactor logging and bioio.

### PR DESCRIPTION
**Pull request by DailyDreaming**
A refactor and attempt to simplify our `bioio` library (which is mostly logging) and some of the utils.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Refactor bioio library, simplify logging, and small adjustments to utils.

### CODE REVIEW

**Review by adamnovak — CHANGES_REQUESTED** [review]
I think the option parser setup should not be in bioio if it needs to know about the existence of provisioners. The provisioner-related options might even want to be defined somewhere in with the provisioners (`toil/provisioners/__init__.py`?) and then pulled over to where the option parser setup lives (`common.py`?).

Other than that this seems mostly good.

File: src/toil/lib/bioio.py [hunk 1]
```diff
@@ -12,292 +12,178 @@
...
-    parser.add_argument("--tempDirRoot", dest="tempDirRoot", type=str,
-                      help="Path to where temporary directory containing all temp files are created, by default uses the current working directory as the base.",
-                      default=tempfile.gettempdir())
+    if not os.path.exists(dirName):
+        os.mkdir(dirName)
+        os.chmod(dirName, 0o777)
+    return dirName
 
+
+def parser_with_common_options(provisioner_options=False):
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/lib/bioio.py:118]: I think that that's a good idea.  Will move it.

I'd also like to remove the `bioio` name and move things to more appropriate places, but we'll need to coordinate with cactus to do so.  I think that the first step should be a deprecation warning and have everything in bioio simply be a wrapper while the real functions are moved out.  I think I'll just go ahead and do that now.

File: src/toil/lib/bioio.py [hunk 2]
```diff
@@ -12,292 +12,178 @@
...
+
+    if provisioner_options:
+        add_provisioner_options(parser)
+
+    # always add these
+    add_logging_options(parser)
+    parser.add_argument("--version", action='version', version=version)
+    parser.add_argument("--tempDirRoot", dest="tempDirRoot", type=str, default=tempfile.gettempdir(),
+                        help="Path to where temporary directory containing all temp files are created, "
+                             "by default uses the current working directory as the base.")
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/lib/bioio.py:129]: Good catch.

File: src/toil/lib/bioio.py [hunk 3]
```diff
@@ -12,292 +12,178 @@
...
+    on. The command can be either a string or a sequence of strings. If it is a string shell=True
+    will be passed to subprocess.check_call.
+    :type command: str | sequence[string]
+    """
+    logger.debug('Running: %r', command)
+    subprocess.check_call(command, shell=isinstance(command, str), bufsize=-1)
+
+
+# used by cactus
+# not used
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/lib/bioio.py:185]: Yes.  Will update to be clearer.

File: src/toil/utils/toilDebugJob.py [hunk 4]
```diff
@@ -11,26 +11,19 @@
...
 
 from toil.common import Config, Toil, jobStoreLocatorHelp
-from toil.lib.bioio import getBasicOptionParser, parseBasicOptions
+from toil.lib.bioio import parser_with_common_options, setLoggingFromOptions
 from toil.utils.toilDebugFile import printContentsOfJobStore
-from toil.version import version
 from toil.worker import workerScript
 
-logger = logging.getLogger( __name__ )
+log = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/utils/toilDebugJob.py:22]: I do.  I'd like to be consistent though.  Either one works for me.

File: src/toil/utils/toilStats.py [hunk 5]
```diff
@@ -573,27 +487,52 @@ def fn4(job):
...
         fileHandle.close()
     # Now dump onto the screen
     print(out_str)
 
+
 def main():
-    """ Reports stats on the workflow, use with --stats option to toil.
-    """
-    parser = getBasicOptionParser()
-    initializeOptions(parser)
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/utils/toilStats.py:594]: Hmmm... I think that that's a reasonable preference and is observed elsewhere.  I'll go ahead and separate it.

File: src/toil/test/utils/utilsTest.py [hunk 6]
```diff
@@ -125,90 +137,93 @@ def testAWSProvisionerUtils(self):
...
-                leader.coreRsync(args=[tmpFile.name, ":"])
-                # Ensure file exists
-                leader.sshAppliance("test", "-e", relpath)
-            tmpDir = tempfile.mkdtemp()
-            # Download the file again and make sure it's the same file
-            # `--protect-args` needed because remote bash chokes on special characters
-            leader.coreRsync(args=["--protect-args", ":" + relpath, tmpDir])
-            with open(os.path.join(tmpDir, relpath), "r") as f:
-                self.assertEqual(f.read(), testData, "Downloaded file does not match original file")
+            # TODO: The following no longer runs programmatically.  Fix and then uncomment.
```
Comment (adamnovak) [inline src/toil/test/utils/utilsTest.py:162]: Are we going to actually fix this? Or should we just cut it?

**Review by adamnovak — CHANGES_REQUESTED** [review]
This looks good, except for this big commented-out test. What are we going to do with it? Do we want to just leave it commented out and merge? It seems like we would want to cut it instead.

File: src/toil/test/utils/utilsTest.py (hunk 6 above)
Comment (DailyDreaming) [reply src/toil/test/utils/utilsTest.py:162]: Hmmm... you raise a good point.  I'll just remove that section then.

**Review by adamnovak — APPROVED** [review]
I think this is good now.

**Pull request merged**